### PR TITLE
GH-37515: [C++] Remove memory address optimization from `ChunkedArray::Equals(const std::shared_ptr<arrow::ChunkedArray>& other)` if the `ChunkedArray` can have `NaN` values

### LINKED ISSUE #37515: [C++] In some cases `ChunkedArray::Equals(std::shared_ptr<arrow::ChunkedArray>& other)` does not treat `NaN` values as unequal consistently
### Describe the bug, including details regarding any error messages, version, and platform.

While working on #37448, I noticed the `Equals()` method of `ChunkedArray` treats NaNs as equal by default. I found this behavior surprising because the `Equals()` method of `Array` doesn't treat NaNs as equal by default.

Below is a program to reproduce this behavior: 

```c++
#include <arrow/api.h>
#include <arrow/type.h>

#include <iostream>
#include <math.h>
#include <sstream>

arrow::Result<std::shared_ptr<arrow::ChunkedArray>> make_chunked_array() {
    arrow::NumericBuilder<arrow::DoubleType> builder;
    
    std::shared_ptr<arrow::Array> array;
    ARROW_RETURN_NOT_OK(builder.AppendValues({0, 1, NAN, 2, 4}));
    ARROW_RETURN_NOT_OK(builder.Finish(&array));
    
    return arrow::ChunkedArray::Make({array});
}

int main(int argc, char *argv[])
{
    auto maybe_chunked_array = make_chunked_array();
    if (!maybe_chunked_array.ok()) {
        return -1;
    }
    auto chunked_array = std::move(maybe_chunked_array).ValueUnsafe();
    auto array = chunked_array->chunk(0);
    
    std::stringstream stream;
    stream << "array contents: ";
    stream << "\n\n";
    stream << array->ToString();
    stream << "\n\n";
    stream << "array->Equals(array): ";
    stream << array->Equals(array);
    stream << "\n\n";

    stream << "chunked_array contents: ";
    stream << "\n\n";
    stream << chunked_array->ToString();
    stream << "\n\n";
    stream << "chunked_array->Equals(chunked_array): ";
    stream << chunked_array->Equals(chunked_array);
    
    std::cout << stream.str() << std::endl;
}
```

Here is the output of this program:

```shell
array contents: 

[
  0,
  1,
  nan,
  2,
  4
]

array->Equals(array): 0

chunked_array contents: 

[
  [
    0,
    1,
    nan,
    2,
    4
  ]
]

chunked_array->Equals(chunked_array): 1
```

Am I correct in that this is a bug?

### Component(s)

C++

### PR DESCRIPTION
**Pull request by sgilmore10**
### Rationale for this change

`ChunkedArray::Equals(const std::shared_ptr<ChunkedArray>& other)` assumes that if the two `ChunkedArray`s share the same memory address, then they must be equal. However, this optimization doesn't take into account that `NaN` values are not considered equal by default. Consequently, this can lead to surprising, inconsistent results from a user's perspective. For example, `ChunkedArray::Equals(const std::shared_ptr<ChunkedArray>& other)`  and `ChunkedArray::Equals(const ChunkedArray& other)` may return different results.

 The program below illustrates this inconsistency:

```c++
#include <arrow/api.h>
#include <arrow/type.h>

#include <iostream>
#include <math.h>
#include <sstream>

arrow::Result<std::shared_ptr<arrow::ChunkedArray>> make_chunked_array() {
    arrow::NumericBuilder<arrow::DoubleType> builder;
    
    std::shared_ptr<arrow::Array> array;
    ARROW_RETURN_NOT_OK(builder.AppendValues({0, 1, NAN, 2, 4}));
    ARROW_RETURN_NOT_OK(builder.Finish(&array));
    
    return arrow::ChunkedArray::Make({array});
}

int main(int argc, char *argv[])
{
    auto maybe_chunked_array = make_chunked_array();
    if (!maybe_chunked_array.ok()) {
        return -1;
    }
    auto chunked_array = std::move(maybe_chunked_array).ValueUnsafe();
    auto array = chunked_array->chunk(0);
    
    std::stringstream stream;

    stream << "chunked_array contents: ";
    stream << "\n\n";
    stream << chunked_array->ToString();
    stream << "\n\n";
    stream << "chunked_array->Equals(chunked_array): ";
    stream << chunked_array->Equals(chunked_array);
    stream << "chunked_array->Equals(*chunked_array): ";
    stream << chunked_array->Equals(*chunked_array);
    
    std::cout << stream.str() << std::endl;
}
```

Here is the output of this program:

```shell
chunked_array contents: 

[
  [
    0,
    1,
    nan,
    2,
    4
  ]
]

chunked_array->Equals(chunked_array): 1
chunked_array->Equals(*chunked_array): 0
```

### What changes are included in this PR?

Updated `ChunkedArray::Equals(const std::shared_ptr<ChunkedArray>& other)` to only return `true` early IF:
   - The two share the same address AND
   - They cannot have `NaN` values 

If both of those conditions are not satisfied, `ChunkedArray::Equals(const std::shared_ptr<ChunkedArray>& other)` will do the element-by-element comparison.

### Are these changes tested?

Yes. I added a new test case called `EqualsSameAddressWithNaNs` to `chunked_array_test.cc`.

### Are there any user-facing changes?

Yes. `ChunkedArray::Equals(const std::shared_ptr<arrow::ChunkedArray>& other)` may return `false` even if the two `ChunkedArray`s have the same memory address. This will only occur if the `ChunkedArray`'s contain `NaN` values.

* Closes: #37515

### CODE REVIEW

File: cpp/src/arrow/chunked_array.cc [hunk 1]
```diff
@@ -111,13 +112,33 @@ bool ChunkedArray::Equals(const ChunkedArray& other) const {
...
+  return supports_nan;
+}
+
+} //  namespace
+
+bool ChunkedArray::Equals(const std::shared_ptr<ChunkedArray>& other) const {
   if (!other) {
     return false;
   }
+  if (this == other.get() && !supportsNaN(*other->type())) {
```
Comment (kou) [inline cpp/src/arrow/chunked_array.cc:139]:
```suggestion
  if (this == other.get() && !supportsNaN(*type_)) {
```

File: cpp/src/arrow/chunked_array_test.cc [hunk 2]
```diff
@@ -146,6 +146,33 @@ TEST_F(TestChunkedArray, EqualsDifferingMetadata) {
   ASSERT_TRUE(left.Equals(right));
 }
 
+TEST_F(TestChunkedArray, EqualsSameAddressWithNaNs) {
+  auto chunk0 = ArrayFromJSON(float64(), "[0, 1, 2, NaN]");
+  auto chunk1 = ArrayFromJSON(float64(), "[3, 4, 5]");
+  ASSERT_OK_AND_ASSIGN(auto result0, ChunkedArray::Make({chunk0, chunk1}, float64()));
+  // result0 has NaN values
+  ASSERT_FALSE(result0->Equals(result0));
```
Comment (kou) [inline cpp/src/arrow/chunked_array_test.cc:154]: Can we use more meaningful names? For example:

```suggestion
  auto chunk_have_nan = ArrayFromJSON(float64(), "[0, 1, 2, NaN]");
  auto chunk_no_nan = ArrayFromJSON(float64(), "[3, 4, 5]");
  ASSERT_OK_AND_ASSIGN(auto chunked_array_have_nan, ChunkedArray::Make({chunk_have_nan, chunk_no_nan}, float64()));
  ASSERT_FALSE(chunked_array_have_nan->Equals(chunked_array_have_nan));
```

**Comment by kou** [comment]
Could you use "what this PR does" instead of "what problem this PR solves" for PR title?

File: cpp/src/arrow/chunked_array.cc [hunk 3]
```diff
@@ -111,13 +112,33 @@ bool ChunkedArray::Equals(const ChunkedArray& other) const {
       .ok();
 }
 
-bool ChunkedArray::Equals(const std::shared_ptr<ChunkedArray>& other) const {
-  if (this == other.get()) {
-    return true;
+namespace {
+
+bool supportsNaN(const arrow::DataType& type) {
```
[the remark this answers is not in the document]
Comment (sgilmore10) [reply cpp/src/arrow/chunked_array.cc:117]: That's definitely a better name. I'll rename the function.

File: cpp/src/arrow/chunked_array_test.cc (hunk 2 above)
Comment (sgilmore10) [reply cpp/src/arrow/chunked_array_test.cc:154]: Will do!

**Comment by sgilmore10** [comment]
> Could you use "what this PR does" instead of "what problem this PR solves" for PR title?

Just renamed the PR to describe what it does instead of the problem it fixes.

**Comment by sgilmore10** [comment]
> Could you fix a lint error by running `ninja format`?
> 
> https://github.com/apache/arrow/actions/runs/6091569576/job/16528361395?pr=37579#step:5:7333
> 
> ```
> --- /arrow/cpp/src/arrow/chunked_array.cc
> +++ /arrow/cpp/src/arrow/chunked_array.cc (after clang format)
> @@ -127,7 +127,7 @@
>    return false;
>  }
>  
> -} //  namespace
> +}  //  namespace
>  
>  bool ChunkedArray::Equals(const std::shared_ptr<ChunkedArray>& other) const {
>    if (!other) {
> ```

Will do!

**Comment by sgilmore10** [comment]
@github-actions autotune

**Comment by sgilmore10** [comment]
I'm not sure why the `continuous-integration/appveyor/pr` workflow failed. I don't think it has anything to do with my changes.

**Review by kou — APPROVED** [review]
+1

The failure on AppVeyor is #37555.

**Pull request merged**